Fix exception, fix hangup at video end, much code cleanup

bepaald opened the pull request:
Ok, this is a big PR, but mostly because I couldn't stand the inconsistent indentation :) And I don't feel like splitting this up in 3 or 4 separate PR's (honestly I wouldn't even know how).

1. It cleans up my own memory leak fix. In this type of program, with a `setup()` and a `loop()`, the things that need to be set once should of course go in the setup, I don't know what I was thinking. This moves creation of the `mp3`, `out` and `aFile` objects to the setup and only calls `aFile::open()` in the player loop when a new file needs to be opened.

   This is really just a clean up, the current method also fixes the leak. The exception you encountered has a different reason. When you suspect a memory leak, adding 
```
  Serial.print(F("Start of loop! Available memory: "));
  Serial.print(esp_get_free_heap_size());
  Serial.println(F(" bytes"));
```
   
   will help. I have done that with this code, and the amount of free memory is stable while looping through a bunch of files.

2. I tried running the full code (without my modifications) to see if I also got the exception you encountered (mentioned here: https://github.com/SuperMakeSomething/mini-video-player/issues/2#issuecomment-1252934705). I _think_ I did (although I can not be 100% sure it's the same problem). This was caused by the mp3 not properly being stopped before starting a new one. I think I fixed this by adding
```
  if (mp3 && mp3->isRunning())
    mp3->stop();
```
   at the start of the player loop. After this I did not encounter any more exceptions.

3. I also experienced the video hanging at the end. Though I saw this behavior every time, since the first code was posted here, so I'm not sure why it worked for you before. This is because the `MjpegClass` was not made to be reused for multiple video's. Some of the values set in it's setup need to be reset when a new file is played. For this reason, this PR adds
```
    _mjpeg_buf_offset = 0;
    _inputindex = 0;
    _remain = 0;
```
   to the MjpegClass. For me, this fixes the hang without the need to stop playback before the file is fully read (this PR also removes the `threshold` code).

4. Removed some unneeded `#include`'s
5. Cleaned up `getNoFiles()` (which seemed to be copy-pasted from arduino's example code without being modified for its new purpose). Also, in my program it always failed to open the files when I set the filenames using `entry.name()`, so this PR changes that to `entry.path()`. Not sure why it would work for you though, so maybe something else is wrong.
6. Changed `Arduino_ST7735` creation and comment to not encourage people to unnecessarily change system wide library files.
7. I think the rest is just changes in whitespace (much needed ;) ).

***NOTE*** I have tested this code, and it seems to work for me. I have no memory leaks, no uncaught exceptions, and no video hanging. But, I do not own any audio hardware, nothing is actually connected on my board for that. So please test this code thoroughly before merging.

Sorry for the big messy PR, I hope it helps you, but feel free to not merge it or pick and choose your changes.

bepaald committed:
Update miniVideoPlayer.ino

bepaald committed:
Update miniVideoPlayer.ino

bepaald committed:
Update MjpegClass.h

SuperMakeSomething commented:
Thank you for all of these contributions! I look forward to testing them. One question — are you using a modified version of `File`? Specifically, I am encountering a compilation error for `entry.path()`. Thank you again for your help!

bepaald commented:
No problem. I think you are probably using an outdated version of the arduino-esp32 board definitions? The `path()` function was introduced in 2.0.0 in april 2021.

In previous versions `name()` did what `path()` does now, which would explain why using `name()` works for you. But they changed `name()` to be compatible with Arduino's original implementation and introduced `path()` to get the old behavior (which is what we need).

Anyway, I think you should update your board definitions. Note, the board manager URL for this has also changed last year: in ArduinoIDE, go to 'file->preferences->settings->additional boards manager URLS' and add `https://raw.githubusercontent.com/espressif/arduino-esp32/gh-pages/package_esp32_index.json` (if it's not there already). Then go to 'tools->board->boards manager', search for 'esp32' and update.

Let me know if this was the issue. Otherwise I could always change the PR back to using name().

SuperMakeSomething commented:
Quick update -- I tried to download the code, but it is resulting in a new set of compilation errors that seem to be related to the ESP8266Audio code compatibility issue with ESP32 libraries newer than 1.0.6.  I noticed this issue before, but punted on it before I could implement a fix for this by making a note in the README.  Would you be willing to change the PR back to using name() instead? I believe that this would allow everything to but still use the older library.  The full error is below.  (Also I noticed a missing `)` on line 246.)  Thanks again for your help!

>In file included from C:\Users\Alex\Documents\Arduino\libraries\ESP8266Audio\src\AudioFileSourceSPIRAMBuffer.h:27,
                 from C:\Users\Alex\Documents\Arduino\libraries\ESP8266Audio\src\AudioFileSourceSPIRAMBuffer.cpp:28:
C:\Users\Alex\Documents\Arduino\libraries\ESP8266Audio\src\spiram-fast.h: In member function 'void ESP8266SPIRAM::begin(int, int)':
C:\Users\Alex\Documents\Arduino\libraries\ESP8266Audio\src\spiram-fast.h:335:26: error: 'SPECIAL' was not declared in this scope
             pinMode(sck, SPECIAL);
                          ^~~~~~~
C:\Users\Alex\Documents\Arduino\libraries\ESP8266Audio\src\spiram-fast.h:335:26: note: suggested alternative: 'SERIAL'
             pinMode(sck, SPECIAL);
                          ^~~~~~~
                          SERIAL
In file included from C:\Users\Alex\Documents\Arduino\libraries\ESP8266Audio\src\AudioGeneratorMIDI.cpp:65:
C:\Users\Alex\Documents\Arduino\libraries\ESP8266Audio\src\libtinysoundfont/tsf.h: In function 'void tsf_channel_midi_control(tsf*, int, int, int)':
C:\Users\Alex\Documents\Arduino\libraries\ESP8266Audio\src\libtinysoundfont/tsf.h:2100:1: error: insn does not satisfy its constraints:
 }
 ^
(insn 858 343 344 51 (set (reg:SF 19 f0 [407])
        (mem/u/c:SF (symbol_ref/u:SI ("*.LC248") [flags 0x2]) [0  S4 A32])) "C:\Users\Alex\Documents\Arduino\libraries\ESP8266Audio\src\libtinysoundfont/tsf.h":2053 47 {movsf_internal}
     (nil))
during RTL pass: postreload
C:\Users\Alex\Documents\Arduino\libraries\ESP8266Audio\src\libtinysoundfont/tsf.h:2100:1: internal compiler error: in extract_constrain_insn, at recog.c:2210
Please submit a full bug report,
with preprocessed source if appropriate.
See <https://gcc.gnu.org/bugs/> for instructions.

bepaald committed:
Update miniVideoPlayer.ino

bepaald commented:
>Quick update -- I tried to download the code, but it is resulting in a new set of compilation errors that seem to be related to the ESP8266Audio code compatibility issue with ESP32 libraries newer than 1.0.6. I noticed this issue before, but punted on it before I could implement a fix for this by making a note in the README. 

This is also because you are running outdated code. ESP8266Audio has also been updated and works fine with newer ESP32 libraries. Why are you actually bundling all the libraries in this project? All of them are available through Arduino IDE's library manager, which will notify you of updates.

>Would you be willing to change the PR back to using name() instead? I believe that this would allow everything to but still use the older library. The full error is below.

Of course, if you really want to. But please do realize that this means the code will not work for anyone who has updated their esp32 board definitions. If you run the old version you _have_ to use `name()`, if you run the new version you _have_ to use `path()`. I'd say it's a better choice to support the current (and future) versions, and not an outdated one, but it's your choice. 

There is a macro available to check the IDF version, but I'm not 100% sure these are available in old version (but I think so). So maybe that will work? Try it out with my last code changes (8876422, already included in this PR).

> (Also I noticed a missing ) on line 246.) Thanks again for your help!

That was already fixed in the second update in this PR (edb499f).

SuperMakeSomething commented:
All I can say is "wow!" Thank you so much for this incredible update.  I've spent some time testing and the code now seems to work without issue.

> Why are you actually bundling all the libraries in this project? 

This is always one of the tricky things I struggle with when I release these projects on YouTube.  In general, I try to make projects as beginner friendly/accessible as possible so that people can have a gentle introduction to Arduino, etc.  From past videos, it seems that newcomers can struggle a bit with library management/having to do too many things at once in order to compile code, which can make them abandon the project.  Bundling libraries with the code seems to be a good middle ground, because it allows them to have everything they need in one package and copying/pasting can be easily explained in the video without it getting too long.  However, I agree that this is definitely not an ideal solution, because the libraries can become outdated.  (I did not expect this to be the case so soon after publishing the video, though I guess that's what I get for using an outdated library during development myself! 😅)  That being said, this code update seems to fix both of those issues by adding the macro to check the IDF version which makes it compatible both with the bundled library as well as with the proper installation method of the latest version through the manager.  I actually had no idea that this was a possibility, so this was a great learning experience for me!

In light of this, I will update the README to reflect these new changes/options and will also add a header to the code to give you credit for your tremendous help!  If we ever meet, I definitely owe you a beer!  Thank you so much!

bepaald commented:
>All I can say is "wow!" Thank you so much for this incredible update. I've spent some time testing and the code now seems to work without issue.

My pleasure

>This is always one of the tricky things I struggle with when I release these projects on YouTube. In general, I try to make projects as beginner friendly/accessible as possible so that people can have a gentle introduction to Arduino, etc. From past videos, it seems that newcomers can struggle a bit with library management/having to do too many things at once in order to compile code, which can make them abandon the project. Bundling libraries with the code seems to be a good middle ground, because it allows them to have everything they need in one package and copying/pasting can be easily explained in the video without it getting too long. However, I agree that this is definitely not an ideal solution, because the libraries can become outdated. (I did not expect this to be the case so soon after publishing the video, though I guess that's what I get for using an outdated library during development myself! sweat_smile) That being said, this code update seems to fix both of those issues by adding the macro to check the IDF version which makes it compatible both with the bundled library as well as with the proper installation method of the latest version through the manager. I actually had no idea that this was a possibility, so this was a great learning experience for me!

Yeah, makes sense.

If I find some time next week, I might write something up for an improved getFilenames() function.
